Basic GPflux interface

sebastianober opened the pull request:
Basic GPflux interface for a "vanilla" DGP model, i.e. using Salimbeni & Deisenroth (2017) inference, no keras layers, no latent variable layers, etc. Simple example of compatible DGP model is `build_constant_input_dim_deep_gp` from `gpflux.architectures`.

TODO/unresolved:
- sampling for whitened posterior representation
- to compile or not compile while sampling in `sample` method?
- avoid retracing with `num_data` for `update` method as in SVGP model interface
- check deepcopy
- better handling of inducing points as dataset grows
- write tests!

Commit:
Minor fixes to error messages

Commit:
Initial attempt at GPflux interface for basic DGP model

Commit:
Fixed warning, update shape check

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,110 @@
+
+
+class GPfluxPredictor(ProbabilisticModel, tf.Module, ABC):
+    """A trainable wrapper for a GPflux deep Gaussian process model. Note that Scipy optimizer
```
note about Scipy is probably not needed here

sebastianober replied:
Done

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,110 @@
+        if optimizer is None:
+            optimizer = TFOptimizer(tf.optimizers.Adam(), batch_size=100)
+
+        if isinstance(optimizer.optimizer, gpflow.optimizers.Scipy):
```
probably there are a lot of optimizers that users might try and it wouldn't work, its probably hopeless to test it this way, to me it seems better to assert we have an instance of TFOptimizer, that class should do lower level checks...

sebastianober replied:
Ok that makes sense, I'll implement that

sebastianober replied:
Done

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,110 @@
+        f_mean, f_var = self.model.predict_f(query_points)
+        return self.model.likelihood_layer.likelihood.predict_mean_and_var(f_mean, f_var)
+
+    def get_kernel(self) -> gpflow.kernels.Kernel:
```
its not a mandatory method so if it doesnt make sense for deep GPs, we can leave it out for now

sebastianober replied:
Done

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,110 @@
+        """
+        self.optimizer.optimize(self.model, dataset)
+
+    __deepcopy__ = module_deepcopy
```
not sure if this is needed... we'll probably need to rethink how we do it, perhaps all model types should implement some kind of `save` method rather than use this, since gpflow, gpflux and keras might do it differently - probably worthwhile raising an issue about this

sebastianober replied:
I have opened issue #338 about this.

hstojic replied:
since we rely on TF probability >= 0.12 this should have been resolved and no need to use current `module_deepcopy` (its also GPflow specific), you might need a different implementation of deepcopy though, we'll see about that

hstojic commented on trieste/models/interfaces.py:
```diff
@@ -97,7 +97,7 @@ def get_kernel(self) -> gpflow.kernels.Kernel:
         :return: The kernel.
         """
-        raise NotImplementedError(f"Model {self!r} does not provide observation noise")
+        raise NotImplementedError(f"Model {self!r} does not provide a kernel")
```
should probably go into a different PR, we don't know when we will merge this one...

hstojic replied:
there was one f string above as well, perhaps a quick PR with these small fixes?

sebastianober replied:
Done, see #336

hstojic commented on trieste/models/optimizer.py:
```diff
@@ -19,8 +19,10 @@
 from typing import Any, Callable, Dict, Iterable, Optional, Tuple, Union
 
 import gpflow
+from gpflux.models import DeepGP
```
right, we should probably have `optimizer.py` file in `models/gpflux/` with gpflux specific stuff

optimizers should have been decomposed alongside the model_interface decomposition but to me it seemed better to leave it for a separate PR

perhaps you can try moving this new code to gpflux folder?

sebastianober replied:
I've had a go at moving the model-specific parts to `optimizer.py` files in the `gpflux` and `gpflow` folders

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,112 @@
+from .interface import GPfluxPredictor
+
+
+class VanillaDeepGP(GPfluxPredictor, TrainableProbabilisticModel):
```
not sure if `Vanilla` is the best name, but well, you do have some artistic freedom here :)

sebastianober replied:
I'm happy to take suggestions, I'm not by any means fixed on Vanilla :)

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,112 @@
+    """
+
+    def __init__(
+        self, model: DeepGP, optimizer: TFOptimizer | None = None
```
if creating a `DeepGP` model is a bit more involved than for the usual `GPflow`, perhaps its worthwhile providing one or two constructors for popular variants? if so, I would do it in a separate file

sebastianober replied:
Yes, I was thinking this, perhaps I could make an `architectures` file or folder in the `gpflux` folder?

hstojic replied:
a single file should do the trick, if you need some supporting functions, put them in a `util.py` file

sebastianober replied:
I've had a go at this for a simple dgp architecture, in `trieste/models/gpflux/architectures.py`

Commit:
Merge branch 'develop' into sebastian/gpflux_dgp

Commit:
Introduced wrapper for basic architecture, moved model-specific
optimizer functions, cleaned files

Commit:
Reformatted

Commit:
Fixed typing, format

Commit:
Fixed types again...

Commit:
And reformat...

sebastianober commented:
Still need to figure out what to do about deepcopy and write tests

Commit:
Hopefully fixed docs?

hstojic commented on trieste/models/gpflow/optimizer.py:
```diff
@@ -0,0 +1,37 @@
+# Copyright 2021 The Trieste Contributors
```
this is good but gpflow changes should be part of a separate PR

a note: `_create_scipy_optimizer` should probably be moved here as well

sebastianober replied:
See #343

hstojic commented on trieste/models/gpflux/optimizer.py:
```diff
@@ -0,0 +1,49 @@
+
+
+@create_loss_function.register
+def _create_loss_function_gpflux(
```
GPflux is using `fit` method from Keras to train DGPs, `fit` provides a lot functionality that we don't want to replicate - `TFOptimizer` will probably work ok, but is not ideal e.g. we create batches there (or through `DatasetTransformer`), I would think Keras' `fit` is more reliable

perhaps we can postpone it for later, but this would require another type of `Optimizer` class, it can be Keras specific or agnostic, such that we can use training method provided by the library, rather than calling an optimizer directly

sebastianober replied:
I have had a go at this, but not by making a new `Optimizer` class, because that seems to run into issues with dispatching to the wrong `create_optimizer` function, as `tf.optimizers.Optimizer` is always a `tf.keras.optimizers.Optimizer`, so that it dispatches to create a `KerasOptimizer` even when we want to create  `TFOptimizer` (for instance, for SVGP). At the moment I don't really see an easy way around this.

My solution was to instead just use the optimizer from `TFOptimizer`, along with the args passed to it, to use `fit` when `model.optimize(Dataset)` is called. See line 98 in `gpflux/interface.py`.

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,107 @@
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        sampler = sample_dgp(self.model)
+
+        @jit(apply=self.optimizer.compile)
```
we can make it an empirical question, make it a parameter in the class and we can test on certain number of datasets and architechtures whether its useful?

sebastianober replied:
Is this referring to whether to compile?

hstojic replied:
yes.... but lets leave this kind of optimization of the code for a later stage

Commit:
Renamed VanillaDeepGP to DeepGaussianProcess

Commit:
Added tests

Commit:
Fixed types

Commit:
Fixed format

hstojic commented on tests/unit/models/gpflux/conftest.py:
```diff
@@ -0,0 +1,24 @@
+import pytest
+
+
+@pytest.fixture(name="compile", params=[True, False])
```
I think there is already this fixture in gpflow conftest, what you should do is delete that one and create conftest in a parent folder and move this fixture there so that both gpflow and gpflux tests use that one

hstojic commented on tests/unit/models/gpflux/test_architectures.py:
```diff
@@ -0,0 +1,86 @@
+# limitations under the License.
+
+"""
+In this module, we test that we are wrapping GPflux architectures correctly, leading to the same
```
looks good!

hstojic commented on tests/util/models/gpflux/models.py:
```diff
@@ -0,0 +1,67 @@
+    return DeepGP([gp_layer_1, gp_layer_2], gpflow.likelihoods.Gaussian(0.01))
+
+
+def fnc_3x_plus_10(x: tf.Tensor) -> tf.Tensor:
```
this function and the one below are the same ones as for gpflow, right? to avoid code duplication lets remove them from gpflow and gpflux and put them in a parent folder

sebastianober replied:
Done

hstojic commented on trieste/models/gpflux/__init__.py:
```diff
@@ -0,0 +1,24 @@
+from .config import GPfluxModelConfig
+from .interface import GPfluxPredictor
+from .models import DeepGaussianProcess
+from .optimizer import _create_loss_function_gpflux
```
hm, not sure this one is needed, I think decorator was supposed to take care of that its available

sebastianober replied:
I've always been a bit confused by this. How would the code work if you don't import it? As in how would Python know to look there if you don't import it?

hstojic commented on tests/unit/models/gpflux/test_models.py:
```diff
@@ -0,0 +1,298 @@
+
+    dgp = DeepGP([gp_layer_1, gp_layer_2], gpflow.likelihoods.Gaussian(0.01))
+
+    with pytest.warns(UserWarning):
```
this recognizes there was a warning, but not what the warning was, right?
you can check for what is contained in the warning message and check if some keyword is present in that string

sebastianober replied:
Yes, that's true, but at the moment we don't have any other `UserWarning` s in the code so hopefully it should be ok? But probably best to be safe about it

hstojic replied:
not at the moment, no, but its easy to add another and forget that this test would then potentially incorrectly pass

sebastianober replied:
Done

hstojic commented on tests/unit/models/gpflux/test_models.py:
```diff
@@ -0,0 +1,298 @@
+    assert model.model.elbo(data) > elbo
+
+
+def _batcher_1(dataset: Dataset, batch_size: int) -> Iterable[tuple[TensorType, TensorType]]:
```
we have these batchers in gpflow tests as well, right? lets move them to a file in `tests/util/models/` folder and load them from there for both gpflow and gpflux

hstojic replied:
or perhaps even better, as a fixture in `tests/unit/models/conftest.py`

sebastianober replied:
Should be done now

Commit:
Swapped over to keras fit for optimization

Commit:
Moved shared fixtures into parent folder

Commit:
added parent file...

Commit:
More specific UserWarning for test_dgp_warns_for_whitened_layers

Commit:
Moved shared util code into parent util directory

Commit:
Reformat

hstojic commented on tests/unit/models/conftest.py:
```diff
@@ -0,0 +1,47 @@
+
+
+@pytest.fixture(name="batcher", params=[_batcher_1, _batcher_2])
+def _dim_fixture(request: Any) -> DatasetTransformer:
```
`dim` is a copy-paste leftover?

sebastianober replied:
Fixed

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,64 @@
+
+from ...types import TensorType
+
+tf.keras.backend.set_floatx("float64")
```
this is important bit but so far the practice was to leave setting floaat type globally to a user, hence we should put it in the tests and notebooks but not here in trieste code

sebastianober replied:
Ok, I will remove this and will add a note in docstrings somewhere so the user is aware as well.

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,112 @@
+class GPfluxPredictor(ProbabilisticModel, tf.Module, ABC):
+    """A trainable wrapper for a GPflux deep Gaussian process model."""
+
+    def __init__(self, optimizer: TFOptimizer | None = None):
```
ok, I would remove dependency on `TFOptimizer` altogether if we are relying on `fit` method, its essentially used for passing around arguments for fit and defining the loss function - instead we can pass the optimizer (`tf.optimizers.Optimizer`) and fit arguments here in init and require models (subclasses) to implement loss method

hstojic replied:
well actually `optimizer` is not needed at all here if we use fit method...

sebastianober replied:
I've had a go at this, let me know what you think

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,112 @@
+
+    @property
+    @abstractmethod
+    def keras_model(self) -> tf.keras.Model:
```
see naming comment in `models.py`

hstojic commented on trieste/models/gpflux/interface.py:
```diff
@@ -0,0 +1,112 @@
+        """
+        self.keras_model.fit(
+            {"inputs": dataset.query_points, "targets": dataset.observations},
+            verbose=0,
```
if we pass fit arguments directly to the class then we can simply unpack the dictionary here, no need for this separated treatment of epochs and batch_size

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,119 @@
+    layers, etc.
+    """
+
+    def __init__(self, model: DeepGP, optimizer: Optimizer | None = None):
```
as in interface, I would pass `tf.optimizers.Optimizer` here and arguments object for `fit` method

sebastianober replied:
This isn't entirely possible this way because of what `create_model_interface` in `ModelConfig` from `models\config.py` expects you to pass to it. I guess one solution would be to write a new `create_model_interface` method for the `GPfluxModelConfig` but that might overcomplicate things?

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,119 @@
+                    "Sampling cannot be currently used with whitening in layers", UserWarning
+                )
+
+        self._model = model
```
perhaps to be clearer, lets have `_model_gpflux` and `_model_keras`, propagated that to the interface as well, and leaving a comment in docstrings somewhere why we have bot versions

sebastianober replied:
Done

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,119 @@
+
+    def __repr__(self) -> str:
+        """"""
+        return f"VanillaDeepGP({self._model!r}, {self.optimizer!r})"
```
`DeepGaussianProcess` no?

sebastianober replied:
Yep, missed that one somehow

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,119 @@
+        self._model = model
+
+        self._keras_model = model.as_training_model()
+        self._keras_model.compile(optimizer.optimizer)
```
shouldn't loss function be used somewhere around here?

sebastianober replied:
I don't think so, with GPflux the loss function is provided automatically, see under `as_training_model` in the API reference: https://secondmind-labs.github.io/GPflux/autoapi/gpflux/models/deep_gp/index.html

hstojic commented on trieste/models/gpflux/optimizer.py:
```diff
@@ -0,0 +1,49 @@
+from ..optimizer import LossClosure, TrainingData, create_loss_function
+
+
+@create_loss_function.register
```
in this new version with `fit` method where is this used exactly?

sebastianober replied:
Forgot to remove this...

Commit:
Save progress

Commit:
Changed way optimizer is used for GPflux models, fixed sampling,
cleaned up various things

Commit:
Formatting

Commit:
Improved testing for sampling

Commit:
Added integration tests

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,161 @@
+
+        super().__init__()
+
+        if isinstance(optimizer, Optimizer):
```
why do you kepp the Optimizer argument?

sebastianober replied:
I got confused and thought the typing wouldn't allow it, but it actually is fine. It's done now

Commit:
Changed GPflux interface to take tf.optimizers.Optimizer instead of
TFOptimizer

Commit:
Format

Commit:
Update to GPflux 0.2.3

Commit:
Whitened layers now supported with sampling (GPflux upgrade)

Commit:
Reset lr on each optimize step in case of lr scheduler

Commit:
Added Michalewicz function, integration test using DGP, Michalewicz,
and Thompson sampling, and removed compile option for DGP sampling
(found to be significantly slower than not compiling)

Commit:
BO integration test

Commit:
Formatting

Commit:
Added tests for an issue encountered with simple GPflux models

Commit:
Initial notebook

Commit:
Actually added notebook...

hstojic commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -0,0 +1,354 @@
+    search_space.upper,
+    grid_density=100
+)
+fig.update_layout(height=400, width=400)
```
fig needs to be bigger in my opinion

hstojic commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -0,0 +1,354 @@
+# %% [markdown]
+# ## Sample the observer over the search space
+#
+# Sometimes we don't have direct access to the objective function. We only have an observer that indirectly observes it. In _Trieste_, an observer can output a number of datasets. In our case, we only have one dataset, the objective. We can convert a function with `scaled_discts_branin`'s signature to a single-output observer using `mk_observer`.
```
you can probably remove this, this is covered in intro notebook with EI
it also has some old refs like `scaled_discts_branin`

hstojic commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -0,0 +1,354 @@
+def build_dgp_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+
+    dgp = build_vanilla_deep_gp(data.query_points, num_layers=2, num_inducing=100)
```
perhaps worth pointing out in the text above that we need inducing points with DGPs

hstojic commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -0,0 +1,354 @@
+    dgp.likelihood_layer.likelihood.variance.assign(1e-5)
+    set_trainable(dgp.likelihood_layer.likelihood.variance, False)
+
+    epochs = 200
```
also worth pointing out we use SGD and Adam to train it, perhaps also mentioning Kereas fit method

hstojic commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -0,0 +1,354 @@
+from util.plotting_plotly import add_bo_points_plotly
+
+fig = plot_function_plotly(function, search_space.lower, search_space.upper, grid_density=100)
+fig.update_layout(height=500, width=500)
```
a bit bigger again, something that would have reasonable size on a usual laptop

hstojic commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -0,0 +1,354 @@
+
+# %% [markdown]
+# ## Describe the problem
+# In this example, we look to find the minimum value of the two-dimensional Branin function over the hypercube $[0, 1]^2$, modified with a discontinuity. We can represent the search space using a `Box`, and plot contours of the Branin over this space.
```
branin? perhaps explain why particular function, why is that an appropriate example for DGPs?

hstojic commented on tests/integration/models/gpflux/test_bayesian_optimization.py:
```diff
@@ -0,0 +1,96 @@
+# Copyright 2021 The Trieste Contributors
```
I think this should go to `tests/integration/test_bayesian_optimization.py` test file rather than specific models folder

hstojic commented on trieste/models/gpflux/models.py:
```diff
@@ -0,0 +1,151 @@
+
+class DeepGaussianProcess(GPfluxPredictor, TrainableProbabilisticModel):
+    """
+    A :class:`TrainableProbabilisticModel` wrapper for a GPflux :class:`~gpflow.models.DeepGP` with
```
gpflow.models?

sebastianober replied:
Should all be done now!

Commit:
Added sample-based expected improvement for DGPs. TODO: tests

Commit:
Cleaned up notebook, fixed some comments

Commit:
Additional comment

Commit:
Moved BO GPflux integration test to parent folder test, formatting

hstojic commented on trieste/objectives/single_objectives.py:
```diff
@@ -383,3 +383,43 @@ def hartmann_6(x: TensorType) -> TensorType:
+
+MICHALEWICZ_10_MINIMUM = tf.constant([-9.66015], tf.float64)
+""" The global minimum of the :func:`michalewicz` function, with shape [1] and dtype float64. """
+
```
you forgot to define search space variables as well, check the other objective functions

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,79 @@ def update_acquisition_function(
+
+        tf.debugging.assert_positive(len(dataset))
+
+        at = dataset.query_points
```
following the suggestion above, lines 246-254 (or 258) would become a method in `DeepGaussianProcess`

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,79 @@ def update_acquisition_function(
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+        sampler = DeepGaussianProcessSampler(self._sample_size, model)
```
following the more general approach, this could be an input to the class instead of being defined here

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,79 @@ def update_acquisition_function(
         return function
 
 
+class DeepGaussianProcessExpectedImprovement(SingleModelAcquisitionBuilder):
```
I would make this more general, not tied specifically to `DeepGaussianProcess` class of models since it could be used by other models (e.g. neural nets) and would remove GPflux dependency here

this can be easily done by providing a sampler as an input and requiring that the `ProbabilisticModel` has to have a method that provides `mean_samples` (we could also have a special subclass of  `ProbabilisticModel` where such method is `@abstractmethod` and then use it here as an input)

sebastianober replied:
I've had a go at this now

hstojic commented on trieste/acquisition/sampler.py:
```diff
@@ -329,6 +330,61 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+    extension of :class:`IndependentReparametrizationSampler` for use with DGP models.
+    """
+
+    def __init__(self, sample_size: int, model: ProbabilisticModel):
```
shouldn't model class here be `DeepGaussianProcess`?

sebastianober replied:
Wouldn't this cause a typing error because we need to init the super class `Sampler`, which takes `ProbabilisticModel`?

hstojic replied:
`DeepGaussianProcess` is a subclass of `ProbabilisticModel` right? hence, it shouldn't because it inherits that class as well

sebastianober replied:
Oh of course

hstojic commented on trieste/acquisition/sampler.py:
```diff
@@ -329,6 +330,61 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+
+            if not eps_is_populated:
+                self._eps_list[i].assign(
+                    tf.random.normal([self._sample_size, tf.shape(mean)[-1]], dtype=tf.float64)
```
not sure I understand why 0,1 parameters here are appropriate?

hstojic commented on trieste/acquisition/sampler.py:
```diff
@@ -329,6 +330,61 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+        if not isinstance(model, DeepGaussianProcess):
+            raise ValueError("Model must be a trieste.models.gpflux.DeepGaussianProcess")
+
+        self._model = model
```
I think this should be provided by the parent class `Sampler`

hstojic commented on trieste/acquisition/sampler.py:
```diff
@@ -329,6 +330,61 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
         return mean[..., None, :, :] + tf.transpose(variance_contribution, new_order)
 
 
+class DeepGaussianProcessSampler(Sampler):
```
it seems there is no simple way to make `IndependentReparametrizationSampler` class more flexible to allow for usage by DGPs - in that case, since this is really DGP specific sampler I would move it to a separate file in `models/gpflux/`

sebastianober replied:
Ok that makes sense

hstojic commented on trieste/acquisition/sampler.py:
```diff
@@ -329,6 +330,61 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+
+        samples = at
+        for i, layer in enumerate(self._model.model_gpflux.f_layers):
+            mean, var = layer.predict(at, full_cov=False, full_output_cov=False)
```
shouldn't here be `samples` instead of `at`?

sebastianober replied:
Yes that's a good catch, thanks!

Commit:
Added search spaces, moved DGP sampler, fixed bug, other minor fixes

Commit:
Made DeepGaussianProcessExpectedImprovement more general, moved
models/gpflux/sampler.py into acquisition folder to avoid circular
imports

Commit:
Removed models/gpflux/sampler

Commit:
Added new sampler in own folder

Commit:
Formatting

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,82 @@ def update_acquisition_function(
+        """
+        :param dataset: The data from the observer. Must be populated.
+        :param model: The model over the specified ``dataset``. Must have event shape [1] and must
+            be an instance of :class:`~trieste.models.gpflux.DeepGaussianProcess`.
```
DeepGaussianProcess

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,82 @@ def update_acquisition_function(
+        :return: The estimated *expected improvement* acquisition function.
+        :raise ValueError (or InvalidArgumentError): If ``dataset`` is not populated, ``model``
+            does not have an event shape of [1] or is not an instance of
+            :class:`~trieste.models.gpflux.DeepGaussianProcess`.
```
DeepGaussianProcess

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,82 @@ def update_acquisition_function(
+            does not have an event shape of [1] or is not an instance of
+            :class:`~trieste.models.gpflux.DeepGaussianProcess`.
+        """
+        mean_samples_op = getattr(model, "mc_posterior_mean", None)
```
this should already throw an exception when the method doesn't exist, but you'll figure out details in tests (this one should be definitely tested)

sebastianober replied:
Do you think it's best to remove this then? It might be nice to have the custom error

hstojic replied:
custom error is good but I'm thinking you need to create condition on this line rather than the one below

sebastianober replied:
I'm not sure I understand what you mean? Which line?

hstojic commented on trieste/acquisition/function.py:
```diff
@@ -201,6 +203,82 @@ def update_acquisition_function(
+            DeepGaussianProcess: DeepGaussianProcessSampler,
+        }
+
+        sampler = None
```
why not simply providing `sampler` in `__init__`?

hstojic commented on trieste/acquisition/models/gpflux/sampler.py:
```diff
@@ -0,0 +1,78 @@
+# Copyright 2021 The Trieste Contributors
```
sorry, what I meant was `trieste/models/gpflux/` not as a subfolder in `acquisition` folder

Commit:
Passing sampler to MCEI, minor fixes

Commit:
Unit tests for function

Commit:
Fixed bug, added tests

Commit:
Added support for latent variable layer in DeepGaussianProcess model.
TODO: tests.

Commit:
Initial interface for global inducing posterior models, including
a useable architecture. TODO: tests

Commit:
Keras integration for global inducing

Commit:
Added plotting for global inducing

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,301 @@
+
+
+def build_vanilla_deep_gp(
+    X: TensorType,
```
for consistency with the rest of the toolbox we should use `query_points` for inputs, instead of `X` here

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,301 @@
+
+
+def build_gi_deep_gp(
+    X: TensorType,
```
`query_points` comment again

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,301 @@
+    X: TensorType,
+    num_layers: int,
+    num_inducing: int,
+    y: Optional[TensorType] = None,
```
`observations` for outputs

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,301 @@
+
+
+def build_latent_variable_dgp_model(
+    X: TensorType,
```
`query_points`

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,301 @@
+        return samples, loss_per_datapoint
+
+
+def _construct_kernel(
```
not sure what to think what to do with this function, why moving it out of the function above?you think you might change it or should be user-specified? perhaps its better left then as an argument to the function?

sebastianober replied:
Yes, I agree it makes more sense in the function, especially since it's specific to the latent variable model. I'll move it.

hstojic commented on trieste/models/gpflux/architectures.py:
```diff
@@ -0,0 +1,301 @@
+    is_first_layer: bool,
+) -> gpflow.kernels.SquaredExponential:
+    if is_first_layer:
+        lengthscales = [0.05] * (input_dim - latent_dim) + [2.0] * latent_dim
```
where do the constants come from?

sebastianober replied:
These are just "sensible" initializations for the model, since we want to have more sensitivity to the data than the latent variables - if they're about the same scale then there's a risk the model will get stuck in a local minimum where it tries to explain all the data as noise.

Commit:
Fixed naming of x, added init from search space

Commit:
Fixed numpy in latent variable sampler

hstojic commented:
@sebastianober we can retire this PR now, right?

hstojic commented:
@sebastianober  we can close this one, right?

sebastianober commented:
Yes, it's good to close!
